fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/digitalcash/TransactionObject.java

```diff
@@ -135,67 +123,30 @@ public boolean isSender(PublicKey publicKey) {
-      throw new IllegalArgumentException(
-          "The public Key is not contained in the receiver public key");
-    }
-    // Set the return value to nothing
-    long miniLao = 0;
-    // Compute the hash of the public key
-    String hashKey = receiver.computeHash();
-    // iterate through the output and sum if it's for the argument public key
-    for (OutputObject outObj : getOutputs()) {
-      if (outObj.getScript().getPubKeyHash().equals(hashKey)) {
-        miniLao += outObj.getValue();
-      }
-    }
-
-    return miniLao;
-  }
-
-  /**
-   * Total MiniLao per public key of a List of Transaction
-   *
-   * @param transactions List<TransactionObject>
-   * @param receiver Public Key
-   * @return long amount per user
-   */
-  public static long getMiniLaoPerReceiverSetTransaction(
-      Set<TransactionObject> transactions, PublicKey receiver) {
-    return transactions.stream().mapToLong(obj -> obj.getMiniLaoPerReceiver(receiver)).sum();
-  }
-
-  /**
-   * Function which return the first amount which correspond to the Public Key (this function is
-   * useful if someone send money to herself, in fact only the first amount in the transaction
-   * correspond to the money he has send to him)
+   * Function that given a Public Key gives the miniLaoCoin received for this transaction object
    *
-   * @param receiver Public Key of a potential receiver
+   * @param user Public Key of a potential receiver
    * @return int amount of Lao Coin
    */
-  public long getMiniLaoPerReceiverFirst(PublicKey receiver) {
-    // Check in the future if useful
-    if (!isReceiver(receiver)) {
-      throw new IllegalArgumentException(
-          "The public Key is not contained in the receiver public key");
-    }
-    // Compute the hash of the public key
-    String computeHash = receiver.computeHash();
-    // iterate through the output and sum if it's for the argument public key
-    for (OutputObject outObj : getOutputs()) {
-      if (outObj.getPubKeyHash().equals(computeHash)) {
-        // return after first occurrence
-        return outObj.getValue();
-      }
-    }
-    return 0;
+  public long getSumForUser(PublicKey user) {
+    String hashKey = user.computeHash();
+
+    // This works because in case of issuance the main pk of the organizer is used as input and not
+    // their token
+    boolean isSender = isSender(user);
+    boolean isIssuance = isCoinBaseTransaction();
+
+    // Iterate over outputs, if issuance -> sum all output where user is
+    // If not, sums all input when user is sender xor recipient (with negative amount in case of
+    // sender)
+    return getOutputs().stream()
+        .filter(
+            outputObject -> {
+              boolean isOwn = outputObject.getPubKeyHash().equals(hashKey);
+              return (isIssuance && isOwn) || !isIssuance && (isSender ^ isOwn);
+            })
+        .mapToLong(outputObject -> isSender ? -outputObject.getValue() : outputObject.getValue())
+        .sum();
```

Comment:
This whole block is not very readable, too much nested logic

Reply:
I've made an alternate solution, tell me what you think 🙂